File: lint_renovate_duplicates.mjs
#!/usr/bin/env node
// lint_renovate_duplicates.mjs
// Lint Renovate JSON5 config for duplicate/overlapping packageRules by matchManagers and matchPackageNames
// Usage: node lint_renovate_duplicates.mjs rules-actions.json5
import fs from 'fs';
import process from 'process';
import json5 from 'json5';

if (process.argv.length < 3) {
  console.error('Usage: node lint_renovate_duplicates.mjs <rules-file.json5>');
  process.exit(1);
}
const path = process.argv[2];
console.log(`[INFO] Linting ${path} for duplicate/overlapping packageRules...`);
let data;
try {
  const raw = fs.readFileSync(path, 'utf8');
  data = json5.parse(raw);
} catch (e) {
  console.error(`[ERROR] Could not parse ${path}: ${e}`);
  process.exit(1);
}
const rules = data.packageRules || [];
console.log(`[INFO] Found ${rules.length} packageRules in ${path}`);
const seen = new Map();
for (let i = 0; i < rules.length; i++) {
  const managers = (rules[i].matchManagers || []).slice().sort().join(',');
  const pkgs = (rules[i].matchPackageNames || []).slice().sort().join(',');
  const key = `${managers}|${pkgs}`;
  if (!seen.has(key)) seen.set(key, []);
  seen.get(key).push(i);
}
for (const [key, idxs] of seen.entries()) {
  if (idxs.length > 1) {
    const [managers, pkgs] = key.split('|');
    console.warn(`[WARN][DUPLICATE] Exact duplicate packageRules in ${path} for matchManagers=[${managers}] and matchPackageNames=[${pkgs}] at indices ${idxs}`);
  }
}
// Warn about rules with same managers and overlapping package names
for (let i = 0; i < rules.length; i++) {
  const m1 = new Set(rules[i].matchManagers || []);
  const p1 = new Set(rules[i].matchPackageNames || []);
  for (let j = i + 1; j < rules.length; j++) {
    const m2 = new Set(rules[j].matchManagers || []);
    const p2 = new Set(rules[j].matchPackageNames || []);
    if (m1.size === m2.size && [...m1].every(x => m2.has(x))) {
      const overlap = [...p1].filter(x => p2.has(x));
      if (overlap.length > 0) {
        console.warn(`[WARN][OVERLAP] Overlapping matchPackageNames in rules ${i} and ${j} for matchManagers=[${[...m1]}]: [${overlap}]`);
      }
    }
  }
}
console.log(`[INFO] Lint complete for ${path}`);

File: lint_renovate_duplicates.py
#!/usr/bin/env python3
"""
Lint Renovate JSON5 config for duplicate matchPackageNames or matchManagers in packageRules.
Warns if any packageRule in a JSON5 file has the same matchManagers and overlapping matchPackageNames as another rule.

Usage:
  python3 lint_renovate_duplicates.py rules-actions.json5
"""
import sys
import re
import json
import json5
from collections import defaultdict

def normalize(val):
    if isinstance(val, list):
        return tuple(sorted(val))
    if isinstance(val, str):
        return val.strip()
    return val

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 lint_renovate_duplicates.py <rules-file.json5>")
        sys.exit(1)
    path = sys.argv[1]
    with open(path) as f:
        raw = f.read()
    try:
        data = json5.loads(raw)
    except Exception as e:
        print(f"ERROR: Could not parse {path}: {e}")
        sys.exit(1)
    rules = data.get('packageRules', [])
    seen = defaultdict(list)
    for idx, rule in enumerate(rules):
        managers = tuple(sorted(rule.get('matchManagers', [])))
        pkgs = tuple(sorted(rule.get('matchPackageNames', [])))
        key = (managers, pkgs)
        seen[key].append(idx)
    # Warn about duplicate rules
    for (managers, pkgs), idxs in seen.items():
        if len(idxs) > 1:
            print(f"WARNING: Duplicate/overlapping packageRules in {path} for matchManagers={managers} and matchPackageNames={pkgs} at indices {idxs}")
    # Warn about rules with same managers and overlapping package names
    for i, rule1 in enumerate(rules):
        m1 = set(rule1.get('matchManagers', []))
        p1 = set(rule1.get('matchPackageNames', []))
        for j, rule2 in enumerate(rules):
            if i >= j:
                continue
            m2 = set(rule2.get('matchManagers', []))
            p2 = set(rule2.get('matchPackageNames', []))
            if m1 == m2 and p1 & p2:
                print(f"WARNING: Overlapping matchPackageNames in rules {i} and {j} for matchManagers={m1}: {p1 & p2}")

if __name__ == "__main__":
    main()
